data_model/src/query/mod.rs

```diff
@@ -31,199 +23,207 @@ use self::{
+/// In the actual implementation, the items collected into batches and a cursor is used to fetch the next batch.
+/// [`builder::IterableQueryIterator`] abstracts over this and allows the query consumer to use a familiar [`Iterator`] interface to iterate over the results.
+pub trait IterableQuery: Sealed {
+    /// The type of single element of the output collection
+    type Item: HasPredicateBox;
 }
 
 #[model]
 mod model {
+
     use getset::Getters;
     use iroha_crypto::HashOf;
-    use iroha_macro::FromVariant;
-    use strum::EnumDiscriminants;
 
     use super::*;
     use crate::block::SignedBlock;
 
-    /// Sized container for all possible Queries.
-    #[allow(clippy::enum_variant_names)]
+    /// An iterable query bundled with a filter
+    ///
+    /// The `P` type doesn't have any bounds to simplify generic trait bounds in some places.
+    /// Use [`super::IterableQueryWithFilterFor`] if you have a concrete query type to avoid specifying `P` manually.
     #[derive(
-        Debug,
-        Display,
-        Clone,
-        PartialEq,
-        Eq,
-        PartialOrd,
-        Ord,
-        EnumRef,
-        EnumDiscriminants,
-        FromVariant,
-        Decode,
-        Encode,
-        Deserialize,
-        Serialize,
-        IntoSchema,
+        Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema, Constructor,
     )]
-    #[enum_ref(derive(Encode, FromVariant))]
-    #[strum_discriminants(
-        vis(pub(super)),
-        name(QueryType),
-        derive(Encode),
-        allow(clippy::enum_variant_names)
+    pub struct IterableQueryWithFilter<Q, P> {
+        pub query: Q,
+        #[serde(default = "predicate_default")]
+        pub predicate: CompoundPredicate<P>,
+    }
+
+    fn predicate_default<P>() -> CompoundPredicate<P> {
+        CompoundPredicate::PASS
+    }
+
+    /// An enum of all possible iterable queries
+    #[derive(
+        Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema, FromVariant,
     )]
-    #[ffi_type]
-    #[allow(missing_docs)]
-    pub enum QueryBox {
-        FindAllAccounts(FindAllAccounts),
-        FindAccountById(FindAccountById),
-        FindAccountKeyValueByIdAndKey(FindAccountKeyValueByIdAndKey),
-        FindAccountsByDomainId(FindAccountsByDomainId),
-        FindAccountsWithAsset(FindAccountsWithAsset),
-        FindAllAssets(FindAllAssets),
-        FindAllAssetsDefinitions(FindAllAssetsDefinitions),
-        FindAssetById(FindAssetById),
-        FindAssetDefinitionById(FindAssetDefinitionById),
-        FindAssetsByName(FindAssetsByName),
-        FindAssetsByAccountId(FindAssetsByAccountId),
-        FindAssetsByAssetDefinitionId(FindAssetsByAssetDefinitionId),
-        FindAssetsByDomainId(FindAssetsByDomainId),
-        FindAssetsByDomainIdAndAssetDefinitionId(FindAssetsByDomainIdAndAssetDefinitionId),
+    pub enum IterableQueryBox {
+        FindAllDomains(IterableQueryWithFilterFor<FindAllDomains>),
```

Comment:
```suggestion
        FindDomains(IterableQueryWithFilterFor<FindAllDomains>),
```

I think that now it makes sense to drop `All` in iterable queries